src/coreclr/jit/gentree.cpp

```diff
@@ -22380,3 +22380,78 @@ bool GenTree::IsInvariant() const
     GenTree* lclVarTree = nullptr;
     return OperIsConst() || Compiler::impIsAddressInLocal(this, &lclVarTree);
 }
+
+//------------------------------------------------------------------------
+// gtReduceStrength: reduce successive variable adds into a single multiply,
+// e.g., i + i + i + i => i * 4.
+//
+// Arguments:
+//    tree - tree for reduction
+//
+// Return Value:
+//    reduced tree if pattern matches, original tree otherwise
+//
+GenTree* Compiler::gtReduceStrength(GenTree* tree)
+{
+    // ADD(_, V0) starts the pattern match
+    if (tree->OperGet() != GT_ADD || tree->gtOverflow() || optValnumCSE_phase)
+    {
+        return tree;
+    }
+
+    genTreeOps oper = tree->OperGet();
+    GenTree*   op1  = tree->AsOp()->gtOp1;
+    GenTree*   op2  = tree->AsOp()->gtOp2;
+
+    if (!op2->OperIs(GT_LCL_VAR) || !varTypeIsIntegralOrI(op2))
+    {
+        return tree;
+    }
+
+    int          foldCount = 0;
+    unsigned int lclNum    = op2->AsLclVarCommon()->GetLclNum();
+
+    // Search for pattern of shape ADD(ADD(ADD(lclNum, lclNum), lclNum), lclNum),
+    while (true)
+    {
+        // ADD(lclNum, lclNum), end of tree
+        if (op1->OperIs(GT_LCL_VAR) && op1->AsLclVarCommon()->GetLclNum() == lclNum && op1->OperIs(GT_LCL_VAR) &&
+            op2->AsLclVarCommon()->GetLclNum() == lclNum)
+        {
+            foldCount += 2;
+            break;
+        }
+        // ADD(ADD(X, Y), lclNum), keep descending
+        else if (op1->OperIs(GT_ADD) && !op1->gtOverflow() && op2->OperIs(GT_LCL_VAR) &&
+                 op2->AsLclVarCommon()->GetLclNum() == lclNum)
+        {
+            foldCount++;
+            op2 = op1->AsOp()->gtOp2;
+            op1 = op1->AsOp()->gtOp1;
+        }
+        // Any other case is a pattern we won't attempt to fold for now
+        else
+        {
+            return tree;
+        }
+    }
+
+    // V0 + V0 ... + V0 becomes V0 * foldCount, where postorder transform will optimize
+    // accordingly
+    GenTree* lclVarTree = tree->AsOp()->gtOp2;
+    GenTree* consTree   = tree->AsOp()->gtOp1;
+    consTree->BashToConst(foldCount, lclVarTree->TypeGet());
+
+    /* GT_MUL operators can later be transformed into 'GT_CALL' */
+    GenTree* morphed =
+        new (this, GT_CALL) GenTreeOp(GT_MUL, tree->TypeGet(), lclVarTree, consTree DEBUGARG(/*largeNode*/ true));
+    if (vnStore != nullptr)
+    {
+        fgValueNumberTreeConst(consTree);
+        fgValueNumberTree(morphed);
```

Comment:
The morphed tree should have the same VN as the original: `morphed->SetVNsFromNode(tree)`. It is not a good idea, in general, to call `fgValueNumberTree` outside of VN, it has not been tested to work that way.

Also, there is a helper you can use here to update the VN of the constant tree: `fgUpdateConstTreeValueNumber`.

Also, if you don't intend to relax the `fgGlobalMorph` check under which this method is right now, you don't need to maintain VNs at all.

Reply:
Thanks for pointing this out. This is mostly my unfamiliarity with the code --- I didn't realize the difference between the `fgGlobalMorph` check, and the `CSE` check, i.e., when the optimizations are valid in which phases.

Looking at the surrounding code in `fgMorphSmpOp`, I see cases where the optimizations are restricted to the global morph, and cases where they are not. Is there a recommended pattern here, or do you have any tips?

Reply:
> Looking at the surrounding code in `fgMorphSmpOp`, I see cases where the optimizations are restricted to the global morph, and cases where they are not. Is there a recommended pattern here, or do you have any tips?

It is mostly cost/benefit question.

But first, legality:
- Only optimizations that preserve value numbers are allowed outside of `fgGlobalMorph`. Allowable VN manipulation is transferring an existing VN from an existing tree, and updating VNs for constants. Anything more complex is probably better left under `fgGlobalMorph`. This optimization can trivially preserve value numbers (the VN of the tree stays the same, as does the VN of the local stays the same, the VN of the constant gets updated).

Allowing optimizations outside of global morph is beneficial because the optimization phases discover new facts that help peepholes, primarily via constant propagation. If your optimization is unlikely to benefit from that, it is likely better to leave it under `fgGlobalMorph` and not worry about VN maintenance / CSEs. Otherwise, it is up to the judgement of the implementer and reviewers what decision is "better" CQ-vs-risk-of-bugs-wise (which, I note, we discover on a somewhat regular basis...).

I personally think it is fine to leave this under `fgGlobalMorph` and delete the VN code. The only new fact that optimizations can discover and which can help the transformation is a local being substituted via copy propagation, but it seems like a rather rare event, and we do not remorph in copy propagation, so it would take a random CSE/assertion to force the remorph and realize the benefit.

(Heh, I should add this my [guide](https://gist.github.com/SingleAccretion/5d18d6de7bdd778543e4ccd4c8825706))